### ISSUE TITLE
Deprecate / remove incomplete calendar view, instead allow export/generation .ics via webcal link

### ISSUE THREAD

**Issue by matthewstrasiotto** [krisives/cronnit.us#60]
Related #16 

Relevant projects: https://github.com/markuspoerschke/iCal

Calendar view is incomplete, & personally I don't have much driving me to finish it- 

There exist a healthy selection of calendar apps (Google, apple, outlook) that are:  
* portable, 
* mature, & 
* feature-packed

whatever we make for cronnit won't be on that level without substantial work, & I just don't see myself using the cronnit calendar view much regardless. 

I think a more useful feature would be to implement an optional feature that allows users to generate a webcal link / "feed" for their posts, which can be integrated into other apps. 

See google calendar's per-calendar "Integrate calendar" setting:

![image](https://user-images.githubusercontent.com/39424834/94393696-f4e45f80-019e-11eb-835a-2b13370b4381.png)

Of course, this comes with a basket of "what-ifs" & security concerns that need to be addressed to ensure this is done safely & privately, as this (read-only) API would probably be independant of the current OAuth based system, instead relying on the secrecy of the calendar link. 

## Security Considerations

I also personally wonder about the privacy of even Google's integration feature, as it seems like the secret lives in the obscurity of the URL, which is susceptible to MITM. 

That said- that's a risk that users assume (& indeed I would accept this risk for the convenience of certain integrations), & google does allow this to be disabled, or for the origical endpoint to be invalidated/reset & a new URL generated.

The hashing / UUID generation needs to be cryptographically secure, & we would need to ensure that it's resistant to rainbow tables, dictionary attacks (salt + pepper will help)

**Comment by MLanghof** [comment]
Some assorted comments:

> I think a more useful feature would be to implement an optional feature that allows users to generate a webcal link / "feed" for their posts, which can be integrated into other apps.

Can you describe what use case you have for this? In my imagination, you are looking to import cronnit-scheduled posts into some calendar app you're already using.

> I also personally wonder about the privacy of even Google's integration feature, as it seems like the secret lives in the obscurity of the URL, which is susceptible to MITM.

How is that HTTPS URL susceptible to MITM?

> The hashing / UUID generation needs to be cryptographically secure, & we would need to ensure that it's resistant to rainbow tables, dictionary attacks, etc (salt & maybe pepper).

To be completely honest, in my only 90% informed opinion, you're listing a bunch of things related to password security here that have no connection to the problem at hand. How do you do dictionary attacks against a UUID?

---

Overall I'm rather skeptical of this, but I'll see what Kris thinks. He might also continue work on the calendar view eventually, so I don't think removing it at this point will be helpful. (We might hide it in the UI for the time being though.)

**Comment by matthewstrasiotto** [comment]
> Can you describe what use case you have for this? In my imagination, you are looking to import cronnit-scheduled posts into some calendar app you're already using.

Yeah pretty much, calendar apps support importing of calendar "feeds" via URL, so, for example, I'm able to share my Uni's timetable with google calendar, and it's dynamically updated. This would provide a read-only view of cronnit posts via the user's choice of calendar app (probably including links in descriptions to edit posts, as well as maybe a synopsis of post body).

> How is that HTTPS URL susceptible to MITM?

NVM - Not sure what I was thinking of but it seems like https urls aren't sent in the clear, besides the DNS stuff. 

> To be completely honest, in my only 90% informed opinion, you're listing a bunch of things related to password security here that have no connection to the problem at hand. How do you do dictionary attacks against a UUID?

I mainly wanted to rattle off a list of concerns regarding the implementation of this feature - that said, i guess the main concern is that I wouldn't want someone to be able to guess this UUID given the name of the account it's for, or even the cronnit account ID. 

I was partially just dumping a bunch of thoughts regarding the feature.

**Comment by matthewstrasiotto** [comment]
I appreciate your skepticism, I'm not set on implementing this - tbh besides my own use cases & interests, I find it hard to gauge what users actually want/need in this app (besides the limited subset that interact w/ the issues here)

**Comment by MLanghof** [comment]
Ok, Kris thinks it's a cool idea, so feel free to take a shot at it if you want! :)

He said:

> To implement each user would get a generated random hash number, or generated based on username, then a URL would exist like: cronnit.us/api/ical?hash=xxx then that would always return a XML iCal document. Users could then copy that URL and use it in their calendar application of choice.

So probably what needs to be done is add an (indexed) column of ical hash to the account table, which we can look up the account ID from. If you want it to be revocable (which seems reasonable enough to me) and non-guessable (which is mandatory), you need some decent source of randomness to hash and a sufficient number of bytes in the hash (i.e. SHA256). Technically the hash is probably redundant, but it shouldn't hurt.

I would like to keep the UI as unobtrusive and lightweight as possible. Not sure where to place the link to it though (and how to do the revocation UI if we want such a thing).

**Comment by matthewstrasiotto** [comment]
Great! Happy to hear Kris is on board

> I would like to keep the UI as unobtrusive and lightweight as possible. Not sure where to place the link to it though 

I think maybe adding it as an item to the view menu, and giving it a full blown page would be ideal?

Or maybe changing the export button to a similar with an "ical feed for use in external calendar app" and "csv" as the options?

(The export menu is probably more cohesive, but the view menu is maybe more discoverable since users are more likely to be clicking there).

> (and how to do the revocation UI if we want such a thing).

I think revocation is important enough to be mandatory, if someone's URL is compromised/leaked, they need some way to unshare their (potentially sensitive) post queue.

If it's given its own page, then that shouldn't be too hard to make look nice- 

I'll probably just take some inspiration from how google calendar does this form, with a preformatted div containing the link, a "copy to clipboard" button, and then some `<sub>` text that explains that its sensitive and can be revoked
